lib/iiif_manifest/v3/manifest_builder/thumbnail_builder.rb

```diff
@@ -0,0 +1,52 @@
+module IIIFManifest
+  module V3
+    class ManifestBuilder
+      class ThumbnailBuilder
+        attr_reader :display_content, :iiif_thumbnail_factory, :image_service_builder_factory
+        def initialize(display_content, iiif_thumbnail_factory:, image_service_builder_factory:)
+          @display_content = display_content
+          @iiif_thumbnail_factory = iiif_thumbnail_factory
+          @image_service_builder_factory = image_service_builder_factory
+        end
+
+        # @return [Array<Object>]
+        def build
+          build_thumbnail
+          image_service_builder.apply(thumbnail)
```

Comment:
I think we need to allow for overriding this behavior for cases where a `iiif_endpoint` isn't present.  For example in avalon we pre-generate thumbnail files and do not run a IIIF image server.  It would be nice if the `display_content` has a `thumbnail` defined to use that (even if nil) and fall back to this helpful default behavior if it isn't defined.  This would also leave open a path for non-image thumbnails or other special cases.  Does that seem workable?

Reply:
I had put [this](https://github.com/samvera/iiif_manifest/blob/1fd685ebf7f204e7186b9713e5fe306edc3b007f/lib/iiif_manifest/v3/manifest_builder/canvas_builder.rb#L65) line in there for cases where there isn't a `iiif_endpoint`, would this be an okay solution for the lack of a `iiif_endpoint`?

As for the pre-generated thumbnail, I think that's a great idea since we don't want to replace it if it's already there.  I am having a hard time visualizing how that looks like.  How does the `display_content` look like coming from Avalon?  Would we need to adjust [this](https://github.com/samvera/iiif_manifest/blob/main/lib/iiif_manifest/v3/display_content.rb) class to accommodate for a thumbnail property?  Something like...?

```ruby
class DisplayContent
  attr_reader :url, :width, :height, :duration, :iiif_endpoint, :format, :type,
              :label, :auth_service, :thumbnail
  def initialize(url, type:, width: nil, height: nil, duration: nil, label: nil,
                 format: nil, iiif_endpoint: nil, auth_service: nil, thumbnail: nil)
    @url = url
    @type = type
    @width = width
    @height = height
    @duration = duration
    @label = label
    @format = format
    @iiif_endpoint = iiif_endpoint
    @auth_service = auth_service
    @thumbnail = thumbnail
  end
end
```

Hmm, but then I guess this won't really work since `thumbnail` is default `nil` so it'd always be set to `nil` instead of building one in the app.  🤔